### ISSUE TITLE
make[3]: *** [/root/ccc/kdns-master/dpdk-19.08/mk/rte.subdir.mk:35: net] Error 2

### ISSUE THREAD

**Issue by hicaoc** [yeze/kdns#1]
centos8.3 

  LD ixgbe_ethdev.o
  AR librte_pmd_ixgbe.a
  INSTALL-LIB librte_pmd_ixgbe.a
  AR librte_pmd_octeontx2.a
  INSTALL-LIB librte_pmd_octeontx2.a
  PMDINFO i40e_ethdev.o.pmd.c
  CC i40e_ethdev.o.pmd.o
  LD i40e_ethdev.o
  AR librte_pmd_i40e.a
  INSTALL-LIB librte_pmd_i40e.a
make[3]: *** [/root/ccc/kdns-master/dpdk-19.08/mk/rte.subdir.mk:35: net] Error 2
make[3]: *** Waiting for unfinished jobs....
  AR librte_pmd_octeontx_crypto.a
  INSTALL-LIB librte_pmd_octeontx_crypto.a
make[2]: *** [/root/ccc/kdns-master/dpdk-19.08/mk/rte.sdkbuild.mk:46: drivers] Error 2
make[1]: *** [/root/ccc/kdns-master/dpdk-19.08/mk/rte.sdkroot.mk:99: all] Error 2
make[1]: Leaving directory '/root/ccc/kdns-master/dpdk-19.08'
make: *** [Makefile:51: dpdk] Error 2
[root@ts2280 kdns-master]#

**Comment by vipinpv85** [comment]
I assume make is run with option `-j` hence we have 2 issues at same place 

1. INSTALL-LIB librte_pmd_i40e.a
2. INSTALL-LIB librte_pmd_octeontx_crypto.a

Looks like file or folder permission issue. Can you please add details on the 

- path it is getting installed?
- permission of the folder you are installing?
- least can you try running with `sudo`?

**Comment by hicaoc** [comment]
make[1]: Entering directory '/root/ccc/kdns-master/core'
  CC dns.o
  CC domain_store.o
  CC packet.o
  CC query.o
  CC radtree.o
  CC util.o
  CC view.o
  CC zone.o
  AR libkdns.a
  INSTALL-LIB libkdns.a
make[1]: Leaving directory '/root/ccc/kdns-master/core'
make[1]: Entering directory '/root/ccc/kdns-master/src'
  CC main.o
/root/ccc/kdns-master/src/main.c: In function ‘set_thread_affinity’:
/root/ccc/kdns-master/src/main.c:126:23: warning: comparison is always true due to limited range of data type [-Wtype-limits]
     for (cid = 0; cid < RTE_MAX_LCORE; ++cid) {
                       ^
/root/ccc/kdns-master/src/main.c:145:23: warning: comparison is always true due to limited range of data type [-Wtype-limits]
     for (cid = 0; cid < RTE_MAX_LCORE; cid++) {
                       ^
/root/ccc/kdns-master/src/main.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-address-of-packed-member’
  CC dns-conf.o
  CC parser.o
  CC netdev.o
  CC forward.o
  CC db_update.o
  CC webserver.o
  CC domain_update.o
  CC view_update.o
  CC kdns-adap.o
  CC tcp_process.o
  CC local_udp_process.o
  CC process.o
  CC hashMap.o
In file included from /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/rte_rwlock.h:12:0,
                 from /root/ccc/kdns-master/src/hashMap.c:1:
/root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:92:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘static’
 static inline int
 ^~~~~~
/root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:136:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘static’
 static inline int
 ^~~~~~
cc1: warning: unrecognized command line option ‘-Wno-address-of-packed-member’
make[2]: *** [/root/ccc/kdns-master/dpdk-19.08/mk/internal/rte.compile-pre.mk:116: hashMap.o] Error 1
make[1]: *** [/root/ccc/kdns-master/dpdk-19.08/mk/rte.extapp.mk:15: all] Error 2
make[1]: Leaving directory '/root/ccc/kdns-master/src'
make: *** [Makefile:62: kdns] Error 2

**Comment by hicaoc** [comment]
> I assume make is run with option `-j` hence we have 2 issues at same place
> 
> 1. INSTALL-LIB librte_pmd_i40e.a
> 2. INSTALL-LIB librte_pmd_octeontx_crypto.a
> 
> Looks like file or folder permission issue. Can you please add details on the
> 
> * path it is getting installed?
> * permission of the folder you are installing?
> * least can you try running with `sudo`?

I'm use root

**Comment by vipinpv85** [comment]
> > I assume make is run with option `-j` hence we have 2 issues at same place
> > 
> > 1. INSTALL-LIB librte_pmd_i40e.a
> > 2. INSTALL-LIB librte_pmd_octeontx_crypto.a
> > 
> > Looks like file or folder permission issue. Can you please add details on the
> > 
> > * path it is getting installed?
> > * permission of the folder you are installing?
> > * least can you try running with `sudo`?
> 
> I'm use root

but from your previous reply, you already some insight to the problem

```
In file included from /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/rte_rwlock.h:12:0,
from /root/ccc/kdns-master/src/hashMap.c:1:
/root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:92:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
static inline int
^~~~~~
/root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:136:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
static inline int
^~~~~~
```

The error you have shared is not coming from DPDK library compilation
```
INSTALL-LIB librte_pmd_i40e.a
make[3]: *** [/root/ccc/kdns-master/dpdk-19.08/mk/rte.subdir.mk:35: net] Error 2
make[3]: *** Waiting for unfinished jobs....
AR librte_pmd_octeontx_crypto.a
INSTALL-LIB librte_pmd_octeontx_crypto.a
make[2]: *** [/root/ccc/kdns-master/dpdk-19.08/mk/rte.sdkbuild.mk:46: drivers] Error 2
make[1]: *** [/root/ccc/kdns-master/dpdk-19.08/mk/rte.sdkroot.mk:99: all] Error 2
```
but from `kdns-master/src/hashMap.c` linking. **Is this a new error?**

**Comment by vipinpv85** [comment]
I think I am to narrow down to your original problem. The project `kdns` has Makefile in the parent Folder, where the expected RTE_TARGET is `x86_64-$(machine)-linuxapp-gcc`. From your logs the RTE_TARGET is set to `arm64-armv8a-linuxapp-gcc`. 

To cross-check if this is the problem, I tried the following

1. git clone the project
2. untar DPDK-19.08
3. cd DPDK folder 
4. export RTE_TARGET=x86_64-native-linuxapp-gcc
5. export RTE_SDK=$PWD
6. make config T=x86_64-native-linuxapp-gcc O=x86_64-native-linuxapp-gcc
7. cd x86_64-native-linuxapp-gcc/
8. make -j (dpdk is built successfully)
9. cd ../../
10. make

I get the dependency errors as 
```bash
/usr/bin/ld: cannot find /tmp/kdns/src/../deps/libmicrohttpd/src/microhttpd/.libs/libmicrohttpd.a: No such file or directory
/usr/bin/ld: cannot find /tmp/kdns/src/../deps/libjansson/src/.libs/libjansson.a: No such file or directory
```

and not

```bash
In file included from /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/rte_rwlock.h:12:0,
from /root/ccc/kdns-master/src/hashMap.c:1:
/root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:92:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
static inline int
^~~~~~
/root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:136:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
static inline int
```

Solution you can try is to edit `Makefile` for your arm target.

**Comment by hicaoc** [comment]
> I think I am to narrow down to your original problem. The project `kdns` has Makefile in the parent Folder, where the expected RTE_TARGET is `x86_64-$(machine)-linuxapp-gcc`. From your logs the RTE_TARGET is set to `arm64-armv8a-linuxapp-gcc`.
> 
> To cross-check if this is the problem, I tried the following
> 
> 1. git clone the project
> 2. untar DPDK-19.08
> 3. cd DPDK folder
> 4. export RTE_TARGET=x86_64-native-linuxapp-gcc
> 5. export RTE_SDK=$PWD
> 6. make config T=x86_64-native-linuxapp-gcc O=x86_64-native-linuxapp-gcc
> 7. cd x86_64-native-linuxapp-gcc/
> 8. make -j (dpdk is built successfully)
> 9. cd ../../
> 10. make
> 
> I get the dependency errors as
> 
> ```shell
> /usr/bin/ld: cannot find /tmp/kdns/src/../deps/libmicrohttpd/src/microhttpd/.libs/libmicrohttpd.a: No such file or directory
> /usr/bin/ld: cannot find /tmp/kdns/src/../deps/libjansson/src/.libs/libjansson.a: No such file or directory
> ```
> 
> and not
> 
> ```shell
> In file included from /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/rte_rwlock.h:12:0,
> from /root/ccc/kdns-master/src/hashMap.c:1:
> /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:92:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
> static inline int
> ^~~~~~
> /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:136:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
> static inline int
> ```
> 
> Solution you can try is to edit `Makefile` for your arm target.

In the x86 platform can be compiled normally through, I this error occurred under the arm64 platform. Then I commented on line 92 and 136 that I could compile through  ：

/root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h

//__rte_experimental     //<-- here 
static inline int
rte_rwlock_read_trylock(rte_rwlock_t *rwl)
{
  `

**Comment by hicaoc** [comment]
> I think I am to narrow down to your original problem. The project `kdns` has Makefile in the parent Folder, where the expected RTE_TARGET is `x86_64-$(machine)-linuxapp-gcc`. From your logs the RTE_TARGET is set to `arm64-armv8a-linuxapp-gcc`.
> 
> To cross-check if this is the problem, I tried the following
> 
> 1. git clone the project
> 2. untar DPDK-19.08
> 3. cd DPDK folder
> 4. export RTE_TARGET=x86_64-native-linuxapp-gcc
> 5. export RTE_SDK=$PWD
> 6. make config T=x86_64-native-linuxapp-gcc O=x86_64-native-linuxapp-gcc
> 7. cd x86_64-native-linuxapp-gcc/
> 8. make -j (dpdk is built successfully)
> 9. cd ../../
> 10. make
> 
> I get the dependency errors as
> 
> ```shell
> /usr/bin/ld: cannot find /tmp/kdns/src/../deps/libmicrohttpd/src/microhttpd/.libs/libmicrohttpd.a: No such file or directory
> /usr/bin/ld: cannot find /tmp/kdns/src/../deps/libjansson/src/.libs/libjansson.a: No such file or directory
> ```
> 
> and not
> 
> ```shell
> In file included from /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/rte_rwlock.h:12:0,
> from /root/ccc/kdns-master/src/hashMap.c:1:
> /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:92:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
> static inline int
> ^~~~~~
> /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:136:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
> static inline int
> ```
> 
> Solution you can try is to edit `Makefile` for your arm target.


you can try :
   make deps && make kdns

**Comment by vipinpv85** [comment]
> arm64-armv8a-linuxapp-gcc

Looks like there is gap here, as per logs you have shared the `RTE_TARGET` set on your build environment is `arm64-armv8a-linuxapp-gcc`. If you are intending to run on `x86 or x86_64` you should not use arm64-armv8a-linuxapp-gcc.

**Comment by vipinpv85** [comment]
> > I think I am to narrow down to your original problem. The project `kdns` has Makefile in the parent Folder, where the expected RTE_TARGET is `x86_64-$(machine)-linuxapp-gcc`. From your logs the RTE_TARGET is set to `arm64-armv8a-linuxapp-gcc`.
> > To cross-check if this is the problem, I tried the following
> > 
> > 1. git clone the project
> > 2. untar DPDK-19.08
> > 3. cd DPDK folder
> > 4. export RTE_TARGET=x86_64-native-linuxapp-gcc
> > 5. export RTE_SDK=$PWD
> > 6. make config T=x86_64-native-linuxapp-gcc O=x86_64-native-linuxapp-gcc
> > 7. cd x86_64-native-linuxapp-gcc/
> > 8. make -j (dpdk is built successfully)
> > 9. cd ../../
> > 10. make
> > 
> > I get the dependency errors as
> > ```shell
> > /usr/bin/ld: cannot find /tmp/kdns/src/../deps/libmicrohttpd/src/microhttpd/.libs/libmicrohttpd.a: No such file or directory
> > /usr/bin/ld: cannot find /tmp/kdns/src/../deps/libjansson/src/.libs/libjansson.a: No such file or directory
> > ```
> > 
> > 
> > and not
> > ```shell
> > In file included from /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/rte_rwlock.h:12:0,
> > from /root/ccc/kdns-master/src/hashMap.c:1:
> > /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:92:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
> > static inline int
> > ^~~~~~
> > /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:136:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
> > static inline int
> > ```
> > 
> > 
> > Solution you can try is to edit `Makefile` for your arm target.
> 
> you can try :
> make deps && make kdns

Thanks for your suggestion, as mentioned above I have no intention to run Kdns and I am not `owner or contributor` for the project. I have replied and tried the steps since it was more of DPDK related setting issue. 

For better help on KDNS reach out to author or contributors. I am done here.

**Comment by hicaoc** [comment]
on arm64 ,I fixed a bug

//set thread's affinity to cpus that are not used by dpdk
static int set_thread_affinity(void) {
    int s;
    uint8_t cid;
    pthread_t tid;
    cpu_set_t cpuset;
    unsigned long long cpumask = 0;

    tid = pthread_self();
    CPU_ZERO(&cpuset);
++    for (cid = 0; cid < RTE_MAX_LCORE-1; ++cid) {
++    for (cid = 0; cid < RTE_MAX_LCORE; ++cid) {
       
            if (!rte_lcore_is_enabled(cid)) {
            CPU_SET(cid, &cpuset);
        }
    }

    s = pthread_setaffinity_np(tid, sizeof(cpu_set_t), &cpuset);
    if (s != 0) {
        log_msg(LOG_ERR, "fail to set thread affinty, errno=%d, errinfo=%s\n", errno, strerror(errno));
        return -1;
    }

    CPU_ZERO(&cpuset);
    s = pthread_getaffinity_np(tid, sizeof(cpu_set_t), &cpuset);
    if (s != 0) {
        log_msg(LOG_ERR, "fail to get thread affinity, errno=%d, errinfo=%s\n", errno, strerror(errno));
        return -2;
    }

  ++  for (cid = 0; cid < RTE_MAX_LCORE-1; cid++) {
   --  for (cid = 0; cid < RTE_MAX_LCORE; cid++) {
        if (CPU_ISSET(cid, &cpuset)) {
            cpumask |= (1LL << cid);
        }
    }

    return 0;
}

**Comment by hicaoc** [comment]
> > > I think I am to narrow down to your original problem. The project `kdns` has Makefile in the parent Folder, where the expected RTE_TARGET is `x86_64-$(machine)-linuxapp-gcc`. From your logs the RTE_TARGET is set to `arm64-armv8a-linuxapp-gcc`.
> > > To cross-check if this is the problem, I tried the following
> > > 
> > > 1. git clone the project
> > > 2. untar DPDK-19.08
> > > 3. cd DPDK folder
> > > 4. export RTE_TARGET=x86_64-native-linuxapp-gcc
> > > 5. export RTE_SDK=$PWD
> > > 6. make config T=x86_64-native-linuxapp-gcc O=x86_64-native-linuxapp-gcc
> > > 7. cd x86_64-native-linuxapp-gcc/
> > > 8. make -j (dpdk is built successfully)
> > > 9. cd ../../
> > > 10. make
> > > 
> > > I get the dependency errors as
> > > ```shell
> > > /usr/bin/ld: cannot find /tmp/kdns/src/../deps/libmicrohttpd/src/microhttpd/.libs/libmicrohttpd.a: No such file or directory
> > > /usr/bin/ld: cannot find /tmp/kdns/src/../deps/libjansson/src/.libs/libjansson.a: No such file or directory
> > > ```
> > > 
> > > 
> > > and not
> > > ```shell
> > > In file included from /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/rte_rwlock.h:12:0,
> > > from /root/ccc/kdns-master/src/hashMap.c:1:
> > > /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:92:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
> > > static inline int
> > > ^~~~~~
> > > /root/ccc/kdns-master/dpdk-19.08/arm64-armv8a-linuxapp-gcc/include/generic/rte_rwlock.h:136:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘attribute’ before ‘static’
> > > static inline int
> > > ```
> > > 
> > > 
> > > Solution you can try is to edit `Makefile` for your arm target.
> > 
> > 
> > you can try :
> > make deps && make kdns
> 
> Thanks for your suggestion, as mentioned above I have no intention to run Kdns and I am not `owner or contributor` for the project. I have replied and tried the steps since it was more of DPDK related setting issue.
> 
> For better help on KDNS reach out to author or contributors. I am done here.

thanks so much 
I've modified the makefile file, and I've found some other issues that have been fixed and are now up and running on the arm platform.



ifdef V
Q =
else
Q = @
endif

ifeq ($(machine),)
**machine = armv8a**
endif

ifeq ($(JOBS),)
    JOBS := $(shell grep -c ^processor /proc/cpuinfo 2>/dev/null)
    ifeq ($(JOBS),)
        JOBS := 1
    endif
endif


#RTE_SDK_TAR = $(CURDIR)/dpdk-19.08.tar.xz
#RTE_SDK = $(CURDIR)/dpdk-19.08
RTE_SDK_TAR = $(CURDIR)/dpdk-19.11.6.tar.xz
RTE_SDK = $(CURDIR)/dpdk-stable-19.11.6
export RTE_SDK


**RTE_TARGET ?= arm64-$(machine)-linuxapp-gcc**
export RTE_TARGET

bindir =  $(CURDIR)/bin
